keps/sig-storage/20200120-skip-permission-change.md

```diff
@@ -0,0 +1,150 @@
+- [Graduation Criteria](#graduation-criteria)
+  - [Test Plan](#test-plan)
+  - [Monitoring](#monitoring)
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
+- [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Summary
+
+Currently before a volume is bind-mounted inside a container the permissions on
+that volume are changed recursively to the provided fsGroup value.  This change
+in ownership can take an excessively long time to complete, especially for very
+large volumes (>=1TB) as well as a few other reasons detailed in [Motivation].
+To solve this issue we will add a new field called `pod.Spec.SecurityContext.FSGroupPermissionChangePolicy` and
+allow the user to specify whether they want the ownership change to occur.
+
+## Motivation
+
+When a volume is mounted on the node, we recursively change permissions of volume
+before bind mounting the volume inside container. The reason of doing this is to ensure
+that volumes are readable/writable by provided fsGroup.
+
+But this presents following problems:
+ - An application(many popular databases) which is sensitive to permission bits changing
+   underneath may refuse to start whenever volume being used inside pod gets mounted on
+   different node.
+ - If volume has a large number of files, performing recursive `chown` and `chmod`
+   could be slow and could cause timeout while starting the pod.
+
+### Goals
+
+ - Allow volume ownership and permission change to be skipped during mount
+
+### Non-Goals
+
+ - In some cases if user brings in a large enough volume from outside, the first time ownership and permission change still could take lot of time.
+ - On SELinux enabled distributions we will still do recursive chcon whenever applicable and handling that is outside the scope.
+ - This proposal does not attempt to fix two pods using same volume with conflicting fsgroup. `FSGroupPermissionChangePolicy` also will be only applicable to volume types which support setting fsgroup.
+
+## Proposal
+
+We propose that an user can optionally opt-in to skip recursive ownership(and permission) change on the volume if volume already has right permissions.
+
+### Implementation Details/Notes/Constraints [optional]
+
+When creating a pod, we propose that `pod.Spec.SecurityContext` field expanded to include a new field called `FSGroupPermissionChangePolicy` which can have following possible values:
+
+ - `Always` --> Always change the permissions and ownership to match fsGroup. This is the current behavior and it will be the default one when this proposal is implemented. Algorithm that performs permission change however will be changed to perform permission change of top level directory last.
+ - `OnDemand` --> Only perform permission and ownership change if permissions of top level directory does not match with expected permissions.
```

Comment:
How does this interact with atomic writer volumes?

Reply:
For atomic writer volumes - the permission change is performed at the end. The only difference is, `ReadOnly` mask gets applied for emptydir/secrets. So permission bits is - `roMask   = os.FileMode(0440)`.

So what will happen is - if root of secret dir(for example) has the correct permissions/ownership already, the permission change will be skipped. IMO - we should be fine.

Reply:
Atomic volumes can have new files added. Will the top level directory still have the old permissions?

Reply:
The permission on the top level directory may actually change on remount call that adds new files to the secret directories - https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/emptydir/empty_dir.go#L352

One way to look at this is - `OnRootDirPermissionMismatch` setting may not work for ephemeral volumes, but technically it is still doing its job. Another thing we could do is to fix it. If `fsGroup` is available in pod's spec then volume's root directory permission need not be 777.  But this is probably a longer discussion . For alpha feature - we could document this as a limitation.

Reply:
Related issue - https://github.com/kubernetes/kubernetes/issues/76158

Reply:
Atomic volumes are max 2MB. I don't think they're going to have the same performance issue that is motivating this change.

I think we actually do want to set permissions every time for atomic volumes since you can have new files added during the remount.